### PR TITLE
SIG Proposal for Sustainability Metrics

### LINKED ISSUE #2020: [WIP] Working Group: Add sustainability Metrics to Semantic Conventions
### Description

With the introduction of various [European Laws regarding sustainability](https://www.consilium.europa.eu/en/policies/green-deal/fit-for-55-the-eu-plan-for-a-green-transition/), it is becoming increasingly more important for companies to track the emissions of their systems.

We have various efforts of implementations in the form of [Kepler](https://github.com/sustainable-computing-io/kepler) and others but creation conventions on tracking these metrics will be vital in order to align with other sectors and remove "creative accounting" practices in the IT sector

We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics accross different ecosystems

### Project Board

Once approved by TC, a project should be managed using a GitHub project board. This project board should be pre-populated with issues that cover all known deliverables, organized by timeline milestones.

A [Technical Committee](https://github.com/open-telemetry/community/blob/main/community-members.md#technical-committee) (TC) member associated with the project can create the board, along with a new project-specific GitHub label to automatically associate issues and PRs with the project. The project lead and all other relevant project members should have edit access to the board.

Once created, please link to the project board here.

### Deliverables

We are looking to add to some new sustainability semantic conventions, these potentially include:

- `hw.abiotic_depletion_potential`
- `hw.product_carbon_footprint`
- `hw.site.pue` ([more](https://knowledge.sdialliance.org/data-center-metrics#903dbdc8aae84211aa42aa33c625ab7e))
- `hw.site.itue`, hw.site.tue ([more](https://knowledge.sdialliance.org/data-center-metrics#e6bcce2b25fb4671a51086dc1eee1695))
- `hw.site.cue` ([more](https://knowledge.sdialliance.org/data-center-metrics#1ae20a03c3f24f6c8558504fdcc6a4af))
- `hw.site.wue`
- `hw.site.ere`, `hw.site.erf` ([more](https://knowledge.sdialliance.org/data-center-metrics#f3b2ee7b318a41a39d16a3cabe2b3ebf))
- `hw.site.ref`, `hw.site.oef` ([more](https://knowledge.sdialliance.org/data-center-metrics#284afeeab1cd474bb0a84dbad125d687))
- `hw.site.cer` ([more](https://knowledge.sdialliance.org/data-center-metrics#fe1467211e0b4f5484f8eae368b4a0a1))
- `hw.site.electricity_cost`
- `hw.site.carbon_intensity`

We will also be creating a prototype based off these

### Staffing / Help Wanted

GC:
- @jpkrohling 

TC:
- none yet

Domain Experts
- @spazzy757
- @gabibeyer
- @bertysentry
- @nikimanoledaki
- @mkorbi
- @rossf7

#### Required staffing

Projects cannot be started until they the following participants have been identified:
* Every project needs a project lead, who is willing to bottom line the project and address any issues which are not handled by other project members.
* At least two sponsoring TC members. TC sponsors are dedicated to attending meetings, reviewing proposals, and in general being aware of the state of the project and it's technical details. TC sponsors guide the project through the spec process, keep the tracking issue up to date, and help to ensure that relevant community members provide input at the appropriate times.
* Engineers willing to write prototypes in at least two languages (if relevant to project). Languages should be fairly different from each other (for example: Java and Python).
* Maintainers or approvers from those languages committed to reviewing the prototypes.

### Meeting Times

- Meeting Times: TBD
- [cncf slack channel](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration

### Timeline

We are hopefully looking at finishing this by end of Q3 2024

### Labels

- `sustainability`

### Linked Issues and PRs
- https://github.com/open-telemetry/semantic-conventions/issues/835

### PR DESCRIPTION
**Pull request by Spazzy757**
This proposes a project and working group for establishing Semantic Conventions for sustainability. 

We aim to unify and and establish standards for monitoring sustainability metrics

Resolves #2020

### CODE REVIEW

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:6]:
```suggestion
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.
```

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:8]:
```suggestion
```

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:9]:
```suggestion
Current efforts of implementations:
```

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:12]: Would people from these projects be willing to be involved here? Do you know if they would be willing to standardize on OTel semantic conventions?

Comment (Spazzy757) [reply projects/env-semconv.md:12]: We can talk with them, and find out, they do have a relationship with TAG ENV

File: projects/env-semconv.md
Comment (trask) [inline projects/env-semconv.md:28]: I confirmed that @pyohannes can be added here as a sponsoring member, thanks @pyohannes!

File: projects/env-semconv.md
Comment (Spazzy757) [reply projects/env-semconv.md:11]: It the Kepler project is interested, I'll push for someone to maybe join

**Comment by almereyda** [comment]
> Resolves https://github.com/open-telemetry/community/pull/2018

This line in the description, which will be in the merge commit, references itself, which is a certain kind of fun in itself, but would probably do good to point at #2020 for people accessing this conversation from the future commit.

File: projects/env-semconv.md
Comment (marceloamaral) [reply projects/env-semconv.md:11]: @pyohannes and @Spazzy757 this could be indeed interesting from the Kepler standpoint

However, some time ago, there was a [presentation during the Kepler community](https://github.com/danielm0hr/edge-metrics-measurements/blob/main/talks/DanielMohr_PromAgentVsOtelCol.pdf) that included an overhead comparison between the OpenTelemetry (OTel) and Prometheus SDKs, showing that the OTel SDK had significantly higher overhead.

As you know, the key difference between these SDKs is how they export metrics. With OpenTelemetry, instrumentation libraries within the application actively push data to backend systems, whereas the Prometheus SDK relies on applications to expose HTTP endpoints for scraping by a monitoring server. These approaches represent orthogonal push and pull approaches.

However, while the Prometheus SDK does offer the option to [enable the OpenMetrics format](https://github.com/prometheus/client_golang/blob/main/examples/exemplars/main.go#L64), it still follows a pull-based approach. 

Given that performance is very important to us, my questions are: do we really need to migrate from Prometheus SDK to OTel SDK? Has the overhead of the OpenTelemetry SDK been improved?

Comment (jgomezselles) [reply projects/env-semconv.md:11]: Hi @marceloamaral, I just wanted to mention that, in those experiments shared in the community, I don't think the OTel SDKs were used. (Please, @danielm0hr, correct me if I'm wrong). Metrics are produced with avalanche in prometheus format and, then, scraped with different mechanisms. (Prometheus, or OpenTelemetry collector in different setups).

My main takeaway from this work is that, when collecting prometheus metrics, (by the time the experiments were made) it was more efficient to just use the prometheus tooling. I don't have proof if this would be the sole contribution for the overhead in usage but my suspicions are pointing to the fact that the OpenTelemetry collector needed to transform from prometheus to OpenTelemetry formats back and forth in this experiment (which is still very valid on its own, but I don't think it will apply for the SDK performance discussion).

For the (very relevant, and important) points that you are raising for assessing the viability of changing the SDK, I think the comparison should be between: 1) metrics generated with the OTel SDK (with the OTel pipeline) vs  2) the current prometheus approach. 

BTW, changing to the OpenTelemetry SDK should not be an impediment to continue using prometheus metrics ( see [prometheus exporter](https://opentelemetry.io/docs/languages/go/exporters/#prometheus-experimental) ).

I hope this helps a bit! We can also follow up in the Kepler community if needed!

Comment (dmathieu) [reply projects/env-semconv.md:11]: The proposal of this issue is also not related to the SDKs, but to agree on a common convention for sustainability-related metrics.

Comment (marceloamaral) [reply projects/env-semconv.md:11]: @jgomezselles, thanks for the clarification! It will indeed be interesting to have another experiment comparing the overhead of the OTel versus Prometheus SDK. Any volunteer? :)

Comment (marceloamaral) [reply projects/env-semconv.md:11]: @dmathieu, right, that makes sense! 
I'd be happy to participate in the discussion.

**Review by jpkrohling — APPROVED** [review]

File: projects/env-semconv.md
Comment (Spazzy757) [reply projects/env-semconv.md:11]: @marceloamaral I have added you to the list of engineers

**Comment by Spazzy757** [comment]
There is a poll for a meeting time in the [CNCF slack channel](https://cloud-native.slack.com/archives/C06RTM63YD6/p1711441500617499) for this project

cc @gabibeyer @bertysentry @mkorbi @rossf7 @marceloamaral

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:37]: This should be added to the `.cspell.yaml`.

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:41]: This can be removed.

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:45]:
```suggestion
- CNCF Slack [#otel-sustainability-semconv-wg](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration
```

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:12]: Do we expect more entries here? Otherwise this could be removed.

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:47]:
```suggestion
Q3 2024

* Kickoff
* Experimental semantic conventions

Q4 2024

* Prototype development

1H 2025

* Review and refinement
* Declare sustainability semantic conventions as stable
```

@Spazzy757 let me know if that makes sense to you. It's better to have a very rough outline and order instead of TBDs. There's nothing cut in stone though, we can adapt as we go.

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:13]:
```suggestion
We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems.
```

**Review by pyohannes — APPROVED** [review]
Thanks @Spazzy757, this looks good.

**Review by trask — APPROVED** [review]

File: projects/env-semconv.md
Comment (danielgblanco) [inline projects/env-semconv.md:19]: I see "recommended practices" is mentioned here, but not in the timeline. Can this be scoped a little tighter? Is this related to "how" to measure or "what/where" to measure for instance? What form will these recommendations be given as?

File: projects/env-semconv.md
Comment (trask) [inline projects/env-semconv.md:41]: (#2044)

```suggestion
- CNCF Slack [#otel-sustainability-semconv-sig](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration
```

**Review by trask — APPROVED** [review]

File: projects/env-semconv.md
Comment (Spazzy757) [reply projects/env-semconv.md:19]: I think this is part of the discussion to be had. I would go to "the recommended way to measure a metric", but think thats up for discussion

File: projects/env-semconv.md
Comment (pyohannes) [inline projects/env-semconv.md:53]:
```suggestion
Q3 2024

* Kickoff
* Experimental semantic conventions

Q4 2024

* Prototype development

1H 2025

* Review and refinement
* Declare sustainability semantic conventions as stable
* Create Recommendations
```

Otherwise it's not formatted correctly.

File: projects/env-semconv.md
Comment (pyohannes) [reply projects/env-semconv.md:19]: Thanks for putting this on the time line.

Given that there are many misconceptions and wrong assumptions around "sustainability", those recommended practices could also offer some guidelines for correctly interpreting those metrics.

File: projects/env-semconv.md
Comment (austinlparker) [inline projects/env-semconv.md:13]: Just a note/quibble I guess, but I think you should be careful with 'metrics' vs. 'Metrics' in the OpenTelemetry sense. For example, there is no reason that various sustainability metrics can't be attached to other data types in OpenTelemetry (e.g., emitting some sort of co2 usage as a span attribute on a given FaaS function execution).

**Review by austinlparker — APPROVED** [review]

File: projects/env-semconv.md
Comment (danielgblanco) [reply projects/env-semconv.md:19]: That makes sense. I just wanted to discuss the potential scope creep in terms of the guidance that OpenTelemetry can give. For instance, what sustainability metrics each organisation should pay more attention to, depending on their workloads and data centre placement, or how to interpret them, is something that other groups are probably better place at answering. However, as that will be discussed as part of this group I'm happy with the answer "within the scope of OpenTelemetry".

**Review by danielgblanco — APPROVED** [review]

**Comment by Spazzy757** [comment]
Hi @pyohannes, is there anything more that this need?

File: projects/env-semconv.md
Comment (danielgblanco) [inline projects/env-semconv.md:27]: I can take this @jpkrohling (there's a requirement for sponsors to be from separate companies).

**Review by svrnm — APPROVED** [review]

**Comment by jpkrohling** [comment]
As discussed during today's GC triage, @danielgblanco will be a GC sponsor, and @pyohannes is the community sponsor delegated by me. I tried to make the required changes to this PR, but it looks like I can't. Here's the proposed diff:

```diff
diff --git a/projects/env-semconv.md b/projects/env-semconv.md
index 7e230e2..58aa81a 100644
--- a/projects/env-semconv.md
+++ b/projects/env-semconv.md
@@ -24,8 +24,8 @@ We want to have a unified conventions in order to better facilitate measuring su
 - @spazzy757
 
 **Sponsoring Members**
-- @jpkrohling 
-- @pyohannes 
+- @danielgblanco
+- @pyohannes (delegated by @jpkrohling, as per [project proposal guidelines](https://github.com/open-telemetry/community/blob/main/project-management.md#project-proposal))
 
 **Engineers**
 - @gabibeyer
 ```

**Comment by jpkrohling** [comment]
There's something else missing here as well: an entry to the sigs.yaml.

https://github.com/open-telemetry/community/blob/main/sigs.yml

@danielgblanco should also be the sponsor there.

**Comment by trask** [comment]
> @pyohannes is the community sponsor delegated by me

semconv maintainers are always allowed to sponsor semconv related SIGs (but looks like we haven't documented it)

**Comment by jpkrohling** [comment]
> semconv maintainers are always allowed to sponsor semconv related SIGs (but looks like we haven't documented it)

I knew I was missing something, and I thought we talked about this somewhere, but couldn't find evidence to back it up :-) Thank you for confirming!

**Comment by danielgblanco** [comment]
@Spazzy757 could you change @jpkrohling for @danielgblanco in the doc? I think this is good to go after that change as we've got the necessary approvals! 🚀

**Comment by Spazzy757** [comment]
Updated!

**Review by danielgblanco — APPROVED** [review]
Looks good! Looking forward to this SIG starting! 💚

**Comment by greenhsu123** [comment]
Hi all, a great proposal. I would like to be added to the list of contributors. Thanks!!

**Comment by danielgblanco** [comment]
As we have a majority of GC votes and the project meets all requirements, I'm merging this. Looking forward to this SIG starting! 🎉  @greenhsu123 can be added later in a different PR (this is standard practice in projects)

@Spazzy757 as lead of this project, can you open an issue in the `community` repo to create meeting calendar, set up necessary GitHub groups, etc? We'll need to add those to the public calendar and list the SIG along with the rest in the official docs.

**Pull request merged**
